### DISCUSSION TITLE
Type inference: Understand that T is always implicitly convertible to T?

### DISCUSSION THREAD

**Discussion by masonwheeler** [dotnet/csharplang#8564] [Language Ideas]
I have a function with this signature:
`public static T? Foo<T>(T? l, T? r) where T : struct, IEqualityOperators<T, T, bool>`

Code that invokes it as `Foo(26, 1953)` gives an error: 
> CS0411: The type arguments for method 'Foo' cannot be inferred from the usage. Try specifying the type arguments explicitly.

I see no ambiguity here.  This should obviously infer the types as `int`.  But per discussion on Discord, the type inference engine doesn't understand that because there are no nullable values being provided, and it's not *quite* smart enough to reason that `T` is always implicitly convertible to `T?`.

Could this be improved in an upcoming C# version?

**Comment by CyrusNajmabadi** [comment]
@masonwheeler 

> Could this be improved in an upcoming C# version?

Yes.  Best way would be to provide a potential spec revision blurb here (you can reference the spec directly, and use things like ```` ```diff ```` sections), stating what change to make to make this work.

**Reply by masonwheeler** [reply]
All right.  Where can I find the spec?

**Reply by 333fred** [reply]
Type inference is specified here: https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/expressions.md#1263-type-inference.

**Reply by masonwheeler** [reply]
Thanks! I'll look it over when I have some time, probably this weekend.
